climada/util/yearsets.py

```diff
@@ -0,0 +1,266 @@
+                                                    nr_annual_events, nr_input_events)
+
+    impact_per_event = np.zeros(np.sum(nr_events))
+    impact_per_year = np.zeros(nr_resampled_years)
+
+    for idx_event, event in enumerate(sampling_vect):
+        impact_per_event[idx_event] = eis.at_event[sampling_vect[event]]
+
+    idx = 0
+    for year in range(nr_resampled_years):
+        impact_per_year[year] = np.sum(impact_per_event[idx:(idx+nr_events[year])])
+        idx += nr_events[year]
+
+    return impact_per_year, nr_events, sampling_vect
+
+
+def sampling_uniform(tot_nr_events, nr_input_events):
+    """Sample uniformely from an array (nr_input_events) for a given amount of
+    events (number events)
+
+    INPUT:
+        tot_nr_events (int): number of events to be resampled
+        nr_input_events (int): number of events contained in given event impact set (eis)
+
+    OUTPUT:
+        sampling_vect (array): sampling vector
+      """
+
+    repetitions = np.ceil(tot_nr_events/(nr_input_events-1)).astype('int')
+
+    rng = default_rng()
+    if repetitions >= 2:
+        sampling_vect = np.round(rng.choice((nr_input_events-1)*repetitions,
+                                            size=tot_nr_events, replace=False)/repetitions
+                                 ).astype('int')
+    else:
+        sampling_vect = rng.choice((nr_input_events-1), size=tot_nr_events,
+                                   replace=False).astype('int')
+
+    return sampling_vect
+
+def sampling_poisson(nr_resampled_years, nr_annual_events, nr_input_events):
+    """Sample amount of events per year following a Poisson distribution
+
+    INPUT:
+        nr_resampled_years (int): the target number of years the impact yearset shall
+            contain.
+        nr_annual_events (int): number of events per year in given event impact set
+        nr_input_events (int): number of events contained in given event impact set (eis)
+
+    OUTPUT:
+        sampling_vect (array): sampling vector
+        nr_events_per_year (array): number of events per resampled year
+    """
+
+    nr_events_per_year = np.round(np.random.poisson(lam=nr_annual_events,
+                                                    size=nr_resampled_years)).astype('int')
+
+    #non_zero_years =  np.where(nr_events_per_year != 0)
+    tot_nr_events = sum(nr_events_per_year)
+
+    sampling_vect = sampling_uniform(tot_nr_events, nr_input_events)
+
+    return sampling_vect, nr_events_per_year
+
+def calculate_correction_fac(impact_per_year, nr_resampled_years, eis):
+    """Apply a correction factor to ensure the expected annual impact (eai) of the annual
+    impact set(ais) amounts to the eai of the event impact set (eis)
+
+    INPUT:
+        impact_per_year (array): resampled annual impact set before applying the correction factor
+        nr_resampled_years (int): the target number of years the annual impact set contains
```

Comment:
Why do you need this, isn't the number of sample years a variable that can be retrieved from eis? The same for impact_per_year?

Reply:
No, both refer to the output of the whole function: the target amount of (re-)sampled years (- maybe n_target_years would be a better name?) and the impact_per_year of the annual impact set

Reply:
But it should be that `len(impact_per_year) == nr_resampled_years` right? This function here is only doing the bias correction, hence the number of years cannot be changed by this function. Thus, the number of years should also not be a necessary input variable.